mypyc/genops.py

```diff
@@ -1163,23 +1362,31 @@ def for_loop_helper(self, index: Lvalue, expr: Expression,
         if (isinstance(expr, CallExpr)
                 and isinstance(expr.callee, RefExpr)
                 and expr.callee.fullname == 'builtins.range'):
+            # TODO: yield statements do not work inside range-based loops.
+
             condition_block = BasicBlock()
             self.push_loop_stack(increment_block, exit_block)
 
             # Special case for x in range(...)
             # TODO: Check argument counts and kinds; check the lvalue
             end = expr.args[0]
-            end_reg = self.accept(end)
 
             # Initialize loop index to 0.
-            index_target = self.get_assignment_target(index)
-            self.assign(index_target, self.accept(IntExpr(0)), line)
+            if self.fn_info.is_generator:
+                assert isinstance(index, NameExpr)
+                assert index.node is not None
+                index_target = self.add_var_to_env_class(index.node, int_rprimitive,
+                                                         self.fn_info.generator_class)
+                self.assign(index_target, self.add(LoadInt(0)), line)
+            else:
+                index_target = self.get_assignment_target(index)
+                self.assign(index_target, self.accept(IntExpr(0)), line)
             self.goto(condition_block)
 
             # Add loop condition check.
             self.activate_block(condition_block)
-            index_reg = self.read(index_target, line)
-            comparison = self.binary_op(index_reg, end_reg, '<', line)
+            end_reg = self.accept(end)
```

Comment:
Oh, oops. I know that I am the one who suggested moving the `end_reg` into the loop but actually it is definitely wrong. Consider `for x in range(foo()): ...`

Reply:
Fixed. I added a test for this as well.